### ISSUE TITLE
Codegen for type-aware allocation attempts to compute the size of an incomplete type when building in c++17 mode

### ISSUE THREAD

**Issue by ojhunt** [llvm/llvm-project#212675]
I've been trying to build a self-targeting fuzzer that does not involve LLMs by working with the various builtin clang AST tooling. Focused solely on operator new because that is something I found bugs in today. Currently it is in no way scalable :D

That aside it found a bug \o/

Prior to c++23 deleting an incomplete object was UB but only produced a warning. In c++23 it became a hard error, but the type-aware allocator logic _assumes_ the c++23 behaviour. As a result building this with clang in c++17 mode this triggers an assertion during codegen:

```cpp
class Foo;
typedef __SIZE_TYPE__ size_t;
namespace std {
  enum class align_val_t : size_t {};
  template <class T> struct type_identity {
    typedef T type;
  };
}
template <class T> void operator delete(std::type_identity<T>, void *, size_t, std::align_val_t);

void f(Foo *o) {
  delete o;
}
```

With libc++ this requires a manual declaration of std::type_identity as it availability gates on the c++ version, I think stdlibc++ does not apply this restriction.

Two possible fixes:
1. Explicitly limit type-aware allocators to >= c++23
2. Reject/don't attempt type aware delete of incomplete types

I think there are reasonable arguments in both directions.

**Comment by ankit-cybertron** [comment]
Hey @ojhunt I'd like to work on this

**Comment by nots1dd** [comment]
> I've been trying to build a self-targeting fuzzer that does not involve LLMs by working with the various builtin clang AST tooling. Focused solely on operator new because that is something I found bugs in today. Currently it is in no way scalable :D
> 
> That aside it found a bug \o/
> 
> Prior to c++23 deleting an incomplete object was UB but only produced a warning. In c++23 it became a hard error, but the type-aware allocator logic _assumes_ the c++23 behaviour. As a result building this with clang in c++17 mode this triggers an assertion during codegen:
> 
> class Foo;
> typedef __SIZE_TYPE__ size_t;
> namespace std {
>   enum class align_val_t : size_t {};
>   template <class T> struct type_identity {
>     typedef T type;
>   };
> }
> template <class T> void operator delete(std::type_identity<T>, void *, size_t, std::align_val_t);
> 
> void f(Foo *o) {
>   delete o;
> }
> 
> With libc++ this requires a manual declaration of std::type_identity as it availability gates on the c++ version, I think stdlibc++ does not apply this restriction.
> 
> Two possible fixes:
> 
>     1. Explicitly limit type-aware allocators to >= c++23
> 
>     2. Reject/don't attempt type aware delete of incomplete types
> 
> 
> I think there are reasonable arguments in both directions.

Hi, just to get a clarification (am not working on this as someone has already chimed in) what are the arguments for not proceeding with limiting the type aware allocators to cpp23? Seems like the most practical thing to do here as option 2 resorts to compiler trying to fix the code that should not compile in the first place (unless am mistaken)

**Comment by linisha15** [comment]
Hey @ojhunt,

I'd like to work on this issue as well. I know @ankit-cybertron is already working on it, but I'm new to LLVM and mainly want to use this as a learning opportunity.

My goal isn't to compete for the issue or have my PR accepted over anyone else's. I just want to work through the implementation, receive feedback on what I did right or wrong, and better understand the project so I can gain the confidence to contribute to more complex issues in the future.

I completely understand that reviewers' time is valuable, and I don't want to create unnecessary duplicate work. That said, would it be okay if I submitted my implementation for review purely for learning purposes, if someone has the time to look at it?

**Comment by ojhunt** [comment]
> Hi, just to get a clarification (am not working on this as someone has already chimed in) what are the arguments for not proceeding with limiting the type aware allocators to cpp23? Seems like the most practical thing to do here as option 2 resorts to compiler trying to fix the code that should not compile in the first place (unless am mistaken)

I think you're right, I'm unsure why I thought the other way would even _work_ let alone be right

**Comment by danielle060821** [comment]
Hi @ankit-cybertron @linisha15,
Are you guys still working on this? If not, I'd be happy to work on this issue

**Comment by ankit-cybertron** [comment]
Hi @danielle060821, thanks for checking! 
Yes, I’m still working on this issue. I have an active PR open as #213455, so I’m waiting for further feedback from the maintainers.
